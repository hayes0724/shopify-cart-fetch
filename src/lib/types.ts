/**
 * The current cart object from Shopify
 * https://shopify.dev/docs/themes/ajax-api/reference/cart#get-cart-js
 */
import {ShopifyError} from "./errors";

export type CartState = {
  readonly token: string;
  readonly note: string;
  readonly attributes: Attributes;
  readonly total_price: number;
  readonly total_weight: number;
  readonly item_count: number;
  readonly items: CartLineItem[];
  readonly requires_shipping: boolean;
  readonly currency: string;
  readonly items_subtotal_price: number;
  readonly original_total_price: number;
  readonly total_discount: number;
  readonly cart_level_discount_applications: CartLevelDiscountApplications[];
};

/**
 * Shopify cart attributes
 * https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-attributes
 */
export type Attributes = {
  [index: string]: string;
};

/**
 * Shopify cart level discounts
 * https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-cart_level_discount_applications
 */
export type CartLevelDiscountApplications = {
  readonly type: string;
  readonly key: string;
  readonly title: string;
  readonly description: string;
  readonly value: string;
  readonly created_at: string;
  readonly value_type: string;
  readonly allocation_method: string;
  readonly target_selection: string;
  readonly target_type: string;
  readonly total_allocated_amount: number;
};

/**
 * Cart line_item object
 * https://shopify.dev/docs/themes/liquid/reference/objects/line_item/
 */
export type CartLineItem = {
  readonly id: number;
  readonly quantity: number;
  readonly variant_id: number;
  readonly key: string;
  readonly title: string;
  readonly price: number;
  readonly line_price: number;
  readonly final_price: number;
  readonly final_line_price: number;
  readonly sku: string;
  readonly grams: number;
  readonly vendor: string;
  readonly taxable: boolean;
  readonly product_id: number;
  readonly product_has_only_default_variant: boolean;
  readonly properties: LineItemProperties;
  readonly gift_card: boolean;
  readonly url: string;
  readonly featured_image: FeaturedImage;
  readonly image: string;
  readonly handle: string;
  readonly requires_shipping: boolean;
  readonly product_title: string;
  readonly product_description: string;
  readonly product_type: string;
  readonly variant_title: string;
  readonly variant_options: string[];
  readonly options_with_values: OptionsWithValues[];
  readonly line_level_discount_allocations?: LineLevelDiscountAllocation[];
};

export type FeaturedImage = {
  readonly url: string;
  readonly aspect_ratio: number;
  readonly alt: string;
};

export type OptionsWithValues = {
  readonly name: string;
  readonly value: string;
};

export type LineLevelDiscountAllocation = {
  amount: number;
  discount_application: DiscountApplication;
};

export type DiscountApplication = {
  type: string;
  key: string;
  title: string;
  description: null;
  value: string;
  created_at: string;
  value_type: string;
  allocation_method: string;
  target_selection: string;
  target_type: string;
  total_allocated_amount: number;
};

export type CartItemAdd = {
  id: VariantID;
  quantity?: number;
  properties?: LineItemProperties;
};

export type CartItemUpdate = {
  quantity: number;
  properties?: LineItemProperties;
} & ({ line: CartItemIndex } | { id: CartItemPropertyID });

export type CartItemRemove = { id: CartItemPropertyID } | { line: CartItemIndex };

export type CartItems = CartItemAdd[] | CartItemAdd

export type CartItemsResponse = { items: CartLineItem[] }

/**
 * The id value is the line item's variant_id or the line item's key.
 */
export type CartItemPropertyID = VariantID | LineItemKey;

/**
 * The 1-based index position of the item in the cart.
 */
export type CartItemIndex = number;

/**
 * The id value is the line item's variant_id or the line item's key.
 * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/line_item#line_item-variant_id | ShopifyAPI: variant_id }
 */
export type VariantID = number | string;

/**
 * The line item's key.
 * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/line_item#line_item-key | ShopifyAPI: line_item.key }
 */
export type LineItemKey = string;

/**
 * An array of custom information for an item that has been added to the cart.
 * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/line_item/#line_item-properties | ShopifyAPI: line_item.properties }
 */
export type LineItemProperties = {
  [index: string]: string;
};

/**
 * Shopify cart API routes
 * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart | ShopifyAPI: cart }
 */
export type CartRoute =
  | "/cart.js"
  | "/cart/add.js"
  | "/cart/update.js"
  | "/cart/change.js"
  | "/cart/clear.js"
  | "/cart/shipping_rates.json"
  | "/cart/prepare_shipping_rates.json"
  | "/cart/async_shipping_rates.json";

export type CartSettings = {
  readonly url?: string;
  readonly postConfig?: RequestInit;
  readonly updateState?: boolean;
};

export type ShopifyResponse = CartState | CartItemsResponse | ShopifyError

export type CartEvents = 'cart:requestStarted' | 'cart:requestComplete' | 'cart:ready'
