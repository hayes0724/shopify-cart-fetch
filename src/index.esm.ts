/**
 * Cart
 * ------------------------------------------------------------------------------
 * The default Cart class with all methods, events and state
 */

import CartEvents from "./lib/events";
import {
  addItem,
  addItemFromForm,
  clearAttributes,
  clearItems,
  clearNote,
  getAttributes,
  getNote,
  getState,
  removeItem,
  updateAttributes,
  updateItem,
  updateItemById,
  updateNote,
} from "./lib/functions";
import {
  CartItemParams,
  CartItemAdd,
  CartItemRemove,
  //CartItemUpdate,
  CartItemIndex,
  CartLineItem,
  CartState,
  EventType,
  LineItemKey,
  LineItemCallback,
  CartItemUpdate,
  CartItemUpdateById,
  Attributes,
  CartParams,
} from "./lib/types";

/**
 * This class implements all methods from functions.ts, uses events.ts for cart events and maintains cart state.
 * You can also import the functions individually or create your own cart class.
 *
 * @example
 * import Cart from '@hayes0724/shopify-cart-fetch';
 * const shopifyCart = new Cart();
 * shopifyCart.getState().then(state => console.log(state))
 * @example
 * import {getState, removeItem, updateItem, addItem, clearItems} from '@hayes0724/shopify-cart-fetch'
 * getState().then(state => console.log(state))
 *
 */
export default class Cart {
  get state(): CartState {
    return this._state;
  }

  set state(value: CartState) {
    this._state = value;
  }

  private _state?: CartState;

  private async run(
    type: EventType,
    callback: CallableFunction,
    data?: CartParams|CartItemParams
  ): Promise<CartState> {
    this.events.emit(type, "start");
    this.state = await callback(data);
    this.events.emit(type, "complete", this.state);
    return this.state;
  }

  private async runLine(
    type: EventType,
    callback: LineItemCallback,
    data?: CartItemParams | CartItemParams[]
  ): Promise<CartLineItem | CartLineItem[]> {
    this.events.emit(type, "start");
    const lineItem = await callback(data);
    const index = this.findItemIndex(lineItem.key);
    index > -1 && lineItem.quantity > 0
      ? (this.state.items[index] = lineItem)
      : this.state.items.push(lineItem);
    this.events.emit(type, "complete", this.state);
    return lineItem;
  }

  public events = new CartEvents();

  /**
   * Finds the the line item using it's line item key
   * @param {LineItemKey} key
   * @return {CartLineItem}
   */
  public findItem = (key: LineItemKey): CartLineItem =>
    this.state.items.find((item) => item.key === key);

  /**
   * Finds the index of the line item using it's line item key
   * @param {LineItemKey} key
   * @return {CartItemIndex}
   */
  public findItemIndex = (key: LineItemKey): CartItemIndex =>
    this.state.items.findIndex((item) => item.key === key);

  public getState = async (): Promise<CartState> =>
    this.run("cart-fetch", getState);

  public addItem = async (items: CartItemAdd[]): Promise<CartLineItem> =>
    this.runLine("cart-add", addItem, items);

  public addItemFromForm = async (
    productForm: HTMLFormElement
  ): Promise<CartLineItem> =>
    this.runLine("cart-add", addItemFromForm, productForm);

  public clearItems = async (): Promise<CartState> =>
    this.run("cart-update", clearItems);

  public removeItem = async (item: CartItemRemove): Promise<CartState> =>
    this.run("cart-remove", removeItem, item);

  public updateItem = async (item: CartItemUpdate): Promise<CartState> =>
    this.run("cart-update", updateItem, item);

  public updateItemById = async (
    item: CartItemUpdateById
  ): Promise<CartLineItem> => this.runLine("cart-update", updateItemById, item);

  public updateAttributes = async (
    attributes: Attributes
  ): Promise<CartState> =>
    this.run("cart-update", updateAttributes, attributes);

  public clearAttributes = async (): Promise<CartState> =>
    this.run("cart-update", clearAttributes);

  public getAttributes: () => Promise<CartState["attributes"]> = getAttributes;

  public getNote: () => Promise<CartState["note"]> = getNote;

  public updateNote = async (note: string): Promise<CartState> =>
    this.run("cart-update", updateNote, note);

  public clearNote = async (): Promise<CartState> =>
    this.run("cart-update", clearNote);
}

export {
  addItem,
  getState,
  addItemFromForm,
  clearItems,
  removeItem,
  updateAttributes,
  updateItem,
  updateItemById,
  getAttributes,
  clearAttributes,
  getNote,
  updateNote,
  clearNote
};

export type {
  CartItemParams,
  CartItemAdd,
  CartItemRemove,
  CartItemIndex,
  CartLineItem,
  CartState,
  EventType,
  LineItemKey,
  LineItemCallback,
  CartItemUpdate,
  CartItemUpdateById,
  Attributes,
  CartParams
};

