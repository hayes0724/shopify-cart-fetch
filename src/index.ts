export * from './shopify-cart';
export * from './lib/types';
export * from './lib/cart-interface';
export * from './lib/errors';
