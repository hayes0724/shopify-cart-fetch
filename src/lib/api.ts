/**
 * Cart API helper methods
 * ------------------------------------------------------------------------------
 * Reusable code for cart API methods
 *
 */

import { CartRoute } from "./types";

/**
 * Creates the default config for post requests
 * @return {JSON}
 */
function defaultRequestConfig() {
  return JSON.parse(
    JSON.stringify({
      method: "POST",
      credentials: 'same-origin',
      headers: {
        'X-Requested-With': 'XMLHttpRequest',
        'Content-Type': 'application/json;'
      }
    })
  );
}

/**
 * The default method for fetching JSON
 * @param {string} route
 * @param {Data?} data
 * @return {Promise<Return>}
 */
async function post<Data, Return>(route: CartRoute, data?: Data): Promise<Return> {
  const config = {
    ...defaultRequestConfig()
  }
  if (data) {
    config.body = JSON.stringify(data)
  }
  return await fetch(route, {...config})
    .then((response) => response.json())
    .then((json: Promise<Return>) => json);
}

export {
  post
}
