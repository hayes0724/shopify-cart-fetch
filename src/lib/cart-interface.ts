import {
  CartState,
  Attributes,
  CartItemRemove,
  CartItemUpdate,
  CartItems,
  CartItemsResponse, CartLineItem
} from "./types";

export interface CartInterface {
  state: CartState
  getState(): Promise<CartState>
  addItem(items: CartItems): Promise<CartItemsResponse>
  addItemFromForm(productForm: HTMLFormElement): Promise<CartLineItem>
  clearItems(): Promise<CartState>
  updateItem(item: CartItemUpdate): Promise<CartState>
  removeItem(item: CartItemRemove): Promise<CartState>
  updateAttributes(attributes: Attributes): Promise<CartState>
  clearAttributes(): Promise<CartState>
  updateNote(note: string): Promise<CartState>
  clearNote(): Promise<CartState>
}
