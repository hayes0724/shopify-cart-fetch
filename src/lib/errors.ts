export interface ShopifyError {
  readonly status: number;
  readonly message: "Cart Error";
  readonly description: string
}

export class InventoryError implements ShopifyError {
  readonly status = 422;
  readonly message = "Cart Error";
  readonly description: string
  constructor(description: string) {
    this.description = description
  }
}

export class VariantError implements ShopifyError {
  readonly status = 404;
  readonly message = "Cart Error";
  readonly description = "Cannot find variant"
}
