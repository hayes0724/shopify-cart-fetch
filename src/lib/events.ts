/**
 * Events
 * ------------------------------------------------------------------------------
 *
 */

import {EventType, EventStage, EventData} from "./types";

export default class CartEvents {

  public bus: EventTarget

  constructor(bus: EventTarget = new DocumentFragment()) {
    this.bus = bus
  }

  static publish(
    type: EventType,
    stage: EventStage,
    data?: CustomEvent["detail"]
  ): CustomEvent<unknown> {
    return new CustomEvent(`${type}:${stage}`, {
      detail: data,
    });
  }

  on(event: string, handler: EventListener): void {
    this.bus.addEventListener(event, handler)
  }

  emit(type: EventType, stage: EventStage, data?: EventData): void {
    this.bus.dispatchEvent(CartEvents.publish(type, stage, data))
  }
}
