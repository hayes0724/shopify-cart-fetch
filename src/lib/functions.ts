import {
  Attributes,
  CartItemAdd,
  CartItemRemove,
  CartItemUpdate,
  CartItemUpdateById,
  CartLineItem,
  CartState
} from "./types";
import {post} from "./api";

/**
 * Use the GET /cart.js endpoint to get the cart as JSON.
 * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#get-cart-js | ShopifyAPI: GET /cart.js }
 * @return {Promise<CartState>}
 */
export async function getState(): Promise<CartState> {
  return await fetch("/cart.js")
    .then((response) => response.json())
    .then((json) => json)
}

/**
 * Use the POST /cart/add.js endpoint to add product to cart.
 * @typedef {function} addItem
 * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-add-js | ShopifyAPI: POST /cart/add.js}
 * @param {CartItemAdd[]} items
 * @return {Promise<CartLineItem[]>}
 */
export async function addItem(items: CartItemAdd[]): Promise<CartLineItem[]> {
  return await post("/cart/add.js", items)
}

/**
 * Adds an item to your cart from a product form. The form must contain an input with name="id".
 * If the quantity specified is more than what is available, the promise will be rejected and the cart state will remain unchanged
 * @param {HTMLFormElement} productForm
 * @return {Promise<CartLineItem>}
 * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-add-js | ShopifyAPI: POST /cart/add.js}
 */
export async function addItemFromForm(
  productForm: HTMLFormElement
): Promise<CartLineItem> {
  const formData = new FormData(productForm);
  if (!formData.get("id")) {
    throw "Cart form missing required property ID";
  }
  return await post("/cart/add.js", formData);
}

/**
 * Use the POST /cart/clear.js endpoint to set all quantities of all line items in the cart to zero.
 * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-clear-js | ShopifyAPI: POST /cart/clear.js}
 * @return {Promise<CartState>}
 */
export async function clearItems(): Promise<CartState> {
  return await post("/cart/clear.js");
}

/**
 * The /cart/change.js endpoint changes the quantity and properties object of a cart line item.
 * Only items already in your cart can be changed, and only one line item at a time can be changed.
 * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-change-js | ShopifyAPI: POST /cart/change.js }
 * @param {CartItemUpdate} item
 * @return {Promise<CartState>}
 */
export async function updateItem(item: CartItemUpdate): Promise<CartState> {
  return await post("/cart/change.js", item);
}

/**
 * The /cart/change.js endpoint changes the quantity.
 * Only items already in your cart can be changed, and only one line item at a time can be changed.
 * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-change-js | ShopifyAPI: POST /cart/change.js }
 * @param {CartItemUpdateById} item
 * @return {Promise<CartLineItem>}
 */
export async function updateItemById(
  item: CartItemUpdateById
): Promise<CartLineItem> {
  return await post("/cart/change.js", item);
}

/**
 * Removes an item from the cart using line item key or product id. Returns a promise which fulfills with the updated cart state.
 * @param {CartItemRemove} item
 * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-change-js | ShopifyAPI: POST /cart/change.js }
 * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/line_item#line_item-key | ShopifyAPI: LineItemKey }
 * @return {Promise<CartState>}
 */
export async function removeItem(item: CartItemRemove): Promise<CartState> {
  return await post("/cart/change.js", { quantity: 0, ...item });
}

/**
 * Get the current cart state from Shopify and return the attributes
 * @return {Promise<CartState["attributes"]>}
 * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-attributes | ShopifyAPI: cart-attributes }
 */
export async function getAttributes(): Promise<CartState["attributes"]> {
  return getState().then((state) => state.attributes);
}

/**
 * Update cart attributes from Shopify and return the state
 * @param {Attributes} attributes
 * @return {Promise<CartState>}
 * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-attributes | ShopifyAPI: cart-attributes }
 */
export async function updateAttributes(
  attributes: Attributes
): Promise<CartState> {
  return await post("/cart/update.js", { attributes: { ...attributes } });
}

/**
 * Clear all cart attributes from Shopify and return the state
 * @return {Promise<CartState>}
 * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-attributes | ShopifyAPI: cart-attributes }
 */
export async function clearAttributes(): Promise<CartState> {
  return getState()
    .then((state) => state.attributes)
    .then((attributes) => {
      const clearList: Attributes = {};
      Object.getOwnPropertyNames(attributes).forEach(
        (prop) => (clearList[prop] = "")
      );
      return clearList;
    })
    .then((clearList) => post("/cart/update.js", { attributes: clearList }));
}

/**
 * Get the current cart state from Shopify and return the note
 * @return {Promise<CartState["note"]>}
 * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-note | ShopifyAPI: cart.note }
 */
export async function getNote(): Promise<CartState["note"]> {
  return getState().then((state) => state.note);
}

/**
 * Update or create the cart note
 * @param {string} note
 * @return {Promise<CartState>}
 * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-note | ShopifyAPI: cart.note }
 */
export async function updateNote(note: string): Promise<CartState> {
  return await post("/cart/update.js", { note: note });
}

/**
 * Remove the cart note
 * @return {Promise<CartState>}
 * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-note | ShopifyAPI: cart.note }
 */
export async function clearNote(): Promise<CartState> {
  return await post("/cart/update.js", { note: "" });
}
