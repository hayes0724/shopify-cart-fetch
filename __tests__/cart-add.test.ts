import { enableFetchMocks } from "jest-fetch-mock";
enableFetchMocks();

import { ShopifyCart } from "../src";
import {
  cartResponse,
  singleItem,
  validForm,
  settings,
  settingsNoUpdate,
  invalidForm,
  multipleItems,
  lineItemResponse,
  inventoryErrorResponse,
  variantErrorResponse,
} from "./fixtures";
import { InventoryError, VariantError } from "../src/lib/errors";

let cart: ShopifyCart;

const verifyResults = (response) => {
  expect(response).toEqual(lineItemResponse);
  expect(cart.state).toEqual(cartResponse);
  expect(fetchMock.mock.calls.length).toEqual(2);
  expect(fetchMock.mock.calls[0][1]["body"]).toMatchSnapshot();
  expect(fetchMock.mock.calls[0][0]).toEqual(`${cart.settings.url}/cart/add.js`);
  expect(fetchMock.mock.calls[1][0]).toEqual(`${cart.settings.url}/cart.js`);
};

const verifyResultsNoUpdate = (response) => {
  expect(response).toEqual(lineItemResponse);
  expect(cart.state).toBeUndefined();
  expect(fetchMock.mock.calls.length).toEqual(1);
  expect(fetchMock.mock.calls[0][1]["body"]).toMatchSnapshot();
  expect(fetchMock.mock.calls[0][0]).toEqual(`${cart.settings.url}/cart/add.js`);
};

describe("Testing methods using the /cart/add.js endpoint and default settings ", () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    cart = new ShopifyCart(settings);
    fetchMock
      .once(JSON.stringify(lineItemResponse))
      .once(JSON.stringify(cartResponse));
  });

  it("addItem calls fetch and updates state", async () => {
    const response = await cart.addItem(singleItem)
    verifyResults(response);
  });

  it("addItem with multiple items calls fetch and updates state", async () => {
    const response = await cart.addItem(multipleItems)
    verifyResults(response);
  });

  it("addItemFromForm calls fetch and updates state", async () => {
    document.body.innerHTML = validForm;
    const form = document.querySelector("form");
    const response = await cart.addItemFromForm(form)
    verifyResults(response);
  });

  it("addItemFromForm throws error when missing id", async () => {
    document.body.innerHTML = invalidForm;
    const form = document.querySelector("form");
    await expect(cart.addItemFromForm(form)).rejects.toMatch(
      "Cart form missing required property ID"
    );
  });
});

describe("Testing methods using the /cart/add.js endpoint without updating state", () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    cart = new ShopifyCart(settingsNoUpdate);
    fetchMock.mockResponseOnce(JSON.stringify(lineItemResponse));
  });

  it("addItem calls fetch without updating state", async () => {
    const response = await cart.addItem(singleItem)
    verifyResultsNoUpdate(response);
  });

  it("addItemFromForm calls fetch without updating state", async () => {
    document.body.innerHTML = validForm;
    const form = document.querySelector("form");
    const response = await cart.addItemFromForm(form)
    verifyResultsNoUpdate(response);
  });
});

describe("Testing API errors using the /cart/add.js endpoint", () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    cart = new ShopifyCart(settings);
  });

  it("addItem calls fetch and throws inventory error", async () => {
    fetchMock.mockResponseOnce(JSON.stringify(inventoryErrorResponse));
    await expect(async () => {
      await cart.addItem(singleItem);
    }).rejects.toBeInstanceOf(InventoryError);
    expect(cart.state).toBeUndefined();
  });

  it("addItem calls fetch and throws variant error", async () => {
    fetchMock.mockResponseOnce(JSON.stringify(variantErrorResponse));
    await expect(async () => {
      await cart.addItem(singleItem);
    }).rejects.toBeInstanceOf(VariantError);
    expect(cart.state).toBeUndefined();
  });
});
