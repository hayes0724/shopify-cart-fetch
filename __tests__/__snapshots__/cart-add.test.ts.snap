// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing methods using the /cart/add.js endpoint and default settings  addItem calls fetch and updates state 1`] = `"{\\"items\\":[{\\"id\\":\\"39766656254012\\"}]}"`;

exports[`Testing methods using the /cart/add.js endpoint and default settings  addItem with multiple items calls fetch and updates state 1`] = `"{\\"items\\":[{\\"id\\":\\"39766656254012\\"},{\\"id\\":\\"39766655926332\\"}]}"`;

exports[`Testing methods using the /cart/add.js endpoint and default settings  addItemFromForm calls fetch and updates state 1`] = `"{\\"id\\":\\"39766656254012\\",\\"quantity\\":\\"1\\"}"`;

exports[`Testing methods using the /cart/add.js endpoint without updating state addItem calls fetch without updating state 1`] = `"{\\"items\\":[{\\"id\\":\\"39766656254012\\"}]}"`;

exports[`Testing methods using the /cart/add.js endpoint without updating state addItemFromForm calls fetch without updating state 1`] = `"{\\"id\\":\\"39766656254012\\",\\"quantity\\":\\"1\\"}"`;
